presto-main/src/main/java/com/facebook/presto/operator/annotations/ScalarImplementationDependency.java

```diff
@@ -15,63 +15,39 @@
 
 import com.facebook.presto.metadata.BoundVariables;
 import com.facebook.presto.metadata.FunctionManager;
+import com.facebook.presto.spi.function.FunctionHandle;
 import com.facebook.presto.spi.function.InvocationConvention;
-import com.facebook.presto.spi.function.Signature;
 import com.facebook.presto.spi.type.TypeManager;
 
 import java.lang.invoke.MethodHandle;
-import java.util.Objects;
 import java.util.Optional;
 
-import static com.facebook.presto.metadata.SignatureBinder.applyBoundVariables;
-import static java.util.Objects.requireNonNull;
-
 public abstract class ScalarImplementationDependency
         implements ImplementationDependency
 {
-    private final Signature signature;
     private final Optional<InvocationConvention> invocationConvention;
 
-    protected ScalarImplementationDependency(Signature signature, Optional<InvocationConvention> invocationConvention)
+    protected ScalarImplementationDependency(Optional<InvocationConvention> invocationConvention)
     {
-        this.signature = requireNonNull(signature, "signature is null");
         this.invocationConvention = invocationConvention;
     }
 
-    public Signature getSignature()
-    {
-        return signature;
-    }
-
+    protected abstract FunctionHandle getFunctionHandle(BoundVariables boundVariables, FunctionManager functionManager);
     @Override
     public MethodHandle resolve(BoundVariables boundVariables, TypeManager typeManager, FunctionManager functionManager)
     {
-        Signature signature = applyBoundVariables(this.signature, boundVariables, this.signature.getArgumentTypes().size());
+        FunctionHandle functionHandle = getFunctionHandle(boundVariables, functionManager);
         if (invocationConvention.isPresent()) {
-            return functionManager.getFunctionInvokerProvider().createFunctionInvoker(signature, invocationConvention).methodHandle();
+            return functionManager.getFunctionInvokerProvider().createFunctionInvoker(functionHandle, invocationConvention).methodHandle();
         }
         else {
-            return functionManager.getScalarFunctionImplementation(signature).getMethodHandle();
+            return functionManager.getScalarFunctionImplementation(functionHandle).getMethodHandle();
         }
     }
 
     @Override
-    public boolean equals(Object o)
-    {
-        if (this == o) {
```

Comment:
Will there be cases where implementations have no fields so they don't have to implement these two methods? Having `abstract` may also be ok to force users to think about this.